src/sshfs_mount/sshfs_mount_handler.cpp

```diff
@@ -60,229 +48,161 @@ void start_and_block_until(mp::Process* process, Signal signal, std::function<bo
+    mpl::log(mpl::Level::info, category, fmt::format("process arguments '{}'", process->arguments().join(", ")));
 
-    start_and_block_until(
-        sshfs_server_process.get(), &mp::Process::ready_read_standard_output, [](mp::Process* process) {
-            return process->read_all_standard_output().contains("Connected"); // Magic string printed by sshfs_server
-        });
+    start_and_block_until_connected(process.get());
 
     // Check in case sshfs_server stopped, usually due to an error
-    auto process_state = sshfs_server_process->process_state();
+    auto process_state = process->process_state();
     if (process_state.exit_code == 9) // Magic number returned by sshfs_server
-    {
-        throw mp::SSHFSMissingError();
-    }
+        throw SSHFSMissingError();
     else if (process_state.exit_code || process_state.error)
-    {
         throw std::runtime_error(
-            fmt::format("{}: {}", process_state.failure_message(), sshfs_server_process->read_all_standard_error()));
-    }
-
-    sshfs_server_configs[vm->vm_name].erase(target_path);
-    mount_processes[vm->vm_name][target_path] = std::move(sshfs_server_process);
+            fmt::format("{}: {}", process_state.failure_message(), process->read_all_standard_error()));
 }
 
-void mp::SSHFSMountHandler::stop_mount(const std::string& instance, const std::string& path)
+void SSHFSMountHandler::stop()
 {
-    auto sshfs_mount_it = mount_processes.find(instance);
-    if (sshfs_mount_it == mount_processes.end())
-    {
-        mpl::log(mpl::Level::info, category,
-                 fmt::format("No running mount process for \"{}\" serving '{}'", instance, path));
+    if (!(process && process->running()))
         return;
-    }
-
-    auto& sshfs_mount_map = sshfs_mount_it->second;
-    auto map_entry = sshfs_mount_map.find(path);
-    if (map_entry != sshfs_mount_map.end())
-    {
-        auto& sshfs_mount = map_entry->second;
-        mpl::log(mpl::Level::info, category, fmt::format("Stopping mount '{}' in instance \"{}\"", path, instance));
-        sshfs_mount->terminate();
 
-        if (!sshfs_mount->wait_for_finished(1000))
-        {
-            mpl::log(mpl::Level::info, category,
-                     fmt::format("Failed to terminate mount '{}' in instance \"{}\", killing", path, instance));
-            sshfs_mount->kill();
-        }
-    }
+    mpl::log(mpl::Level::info, category, fmt::format("Stopping mount \"{}\" in instance '{}'", target, vm->vm_name));
+    if (process->terminate(); !process->wait_for_finished(5000))
+        throw std::runtime_error(
+            fmt::format("Failed to terminate SSHFS mount process: {}", process->read_all_standard_error()));
 }
-
-void mp::SSHFSMountHandler::stop_all_mounts_for_instance(const std::string& instance)
+SSHFSMountHandler::~SSHFSMountHandler()
 {
-    auto mounts_it = mount_processes.find(instance);
-    if (mounts_it == mount_processes.end() || mounts_it->second.empty())
+    try
```

Comment:
It kind of is common, one difference being the `category` used in logging. But still, pulling it to the base class wouldn't work, since calling a virtual method inside a destructor bypasses virtual dispatch, so the base destructor wouldn't call the `stop` function of the derived class.

Reply:
Very good point. What you could do is have a regular protected function in the base that you'd call explicitly in each derived class's destructor. But perhaps not worth the trouble on this occasion.